### PR TITLE
Fix excessive messaging during restart (+ memory issues).

### PR DESCRIPTION
**Pull request by liayoo**
- Fixed node restarting issues (excessive messaging, memory increasing & dying, getting stuck in "needs syncing", ...)
  - Changed to get chain segments from 1 randomly selected peer
  - The chain sync peer gets updated if it gives me an invalid chain / disconnects / I'm fully synced with it
  - Updated to not send duplicate chain segment request messages
- Added a node state check in executeAndBroadcastTransaction()
- Added a consensus state check before sending consensus msgs to peers

Related issues: #432 #510 #562 #653 #695

### CODE REVIEW

**Comment by liayoo** [comment]
Restarting with 2K blocks (full-sync)
=> Success
Screen Shot 2021-11-23 at 9 44 50 PM
Screen Shot 2021-11-23 at 9 55 09 PM

Restarting with  20K blocks (full-sync)
=> Success
Screen Shot 2021-11-24 at 12 26 39 PM
Screen Shot 2021-11-24 at 12 26 59 PM


\+ fast syncing worked as well for both cases
\+ discovered that memory usage comes down quite slowly, after a rapid increase during the sync
\+ we should look into further optimizing & reducing memory usage during restart w/ many blocks..

File: p2p/index.js [hunk 1]
```diff
@@ -215,6 +216,30 @@ class P2pClient {
     }
   }
 
+  /**
+   * Use the existing peer or, if a peer is unavailable, randomly assign a peer for syncing the chain.
```
Comment (platfowner) [inline p2p/index.js:220]: s/a  peer is/the peer is/?

File: p2p/index.js [hunk 2]
```diff
@@ -320,23 +343,43 @@ class P2pClient {
...
+      return;
+    }
     const payload = encapsulateMessage(MessageTypes.CHAIN_SEGMENT_REQUEST, { lastBlockNumber });
     if (!payload) {
-      logger.error('The request chainSegment cannot be sent because of msg encapsulation failure.');
+      logger.error(`[${LOG_HEADER}] The request chainSegment cannot be sent because of msg encapsulation failure.`);
       return;
     }
+    this.chainSyncInProgress.lastBlockNumber = lastBlockNumber;
+    this.chainSyncInProgress.updatedAt = Date.now();
```
Comment (platfowner) [inline p2p/index.js:382]: minor: what about adding a separate function for the above updates? e.g. this.updateChainSyncPeer(lastBlockNumber)?

File: p2p/server.js [hunk 3]
```diff
@@ -580,6 +576,7 @@ class P2pServer {
             if (this.node.state !== BlockchainNodeStates.SERVING) {
               logger.debug(`[${LOG_HEADER}] Not ready to accept chain segment request.\n` +
                   `My node status is now ${this.node.state}.`);
+              this.client.requestChainSegment();
```
Comment (platfowner) [inline p2p/server.js:579]: 👍

File: p2p/index.js [hunk 4]
```diff
@@ -516,33 +562,40 @@ class P2pClient {
 
   handleChainSegment(number, chainSegment, catchUpInfo, socket) {
     const LOG_HEADER = 'handleChainSegment';
+    const address = getAddressFromSocket(this.outbound, socket);
+    if (_.get(this.chainSyncInProgress, 'address') !== address) { // Received from a peer that I didn't request from
```
Comment (platfowner) [inline p2p/index.js:566]: 👍

File: p2p/index.js [hunk 5]
```diff
@@ -320,23 +343,43 @@ class P2pClient {
...
       return;
     }
+    const socket = this.assignRandomPeerForChainSync();
+    if (!socket) {
+      logger.error(`[${LOG_HEADER}] Failed to get a peer for CHAIN_SEGMENT_REQUEST`);
+      return;
+    }
     const lastBlockNumber = this.server.node.bc.lastBlockNumber();
+    if (this.chainSyncInProgress.lastBlockNumber >= lastBlockNumber &&
+        this.chainSyncInProgress.updatedAt > Date.now() - BlockchainConfigs.EPOCH_MS) { // time buffer
```
Comment (platfowner) [inline p2p/index.js:372]: nice approach!

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

**Comment by platfowner** [comment]
And thanks for sharing the insights.

File: p2p/index.js (hunk 2 above)
Comment (liayoo) [reply p2p/index.js]: Sounds good! Updated.

File: p2p/index.js (hunk 1 above)
Comment (liayoo) [reply p2p/index.js]: Done!

**Comment by liayoo** [comment]
Thanks for the review!! Now merging..

**Pull request merged**